Data Types - Add Nested Fields Support

gmourier opened the pull request:
The definitive specification is due at **2022-04-01**

---

# Summary

Lists the different data types supported by Meilisearch and adds a specific section for nested structures. 

Since the fields are recursively flattened at indexing time internally, it makes it possible for a user to express a nested field location in any API parameter accepting document attributes using the `.` (dot-notation).

Example: 

```json
{
    "id": 0,
    "person": {
         "firstname": "Guillaume"
    }
}
```

Is internally represented

```json
{
    "id": 0,
    "person.firstname": "Guillaume"
}
```

It can be expressed this way in any API parameter accepting document attributes. 

e.g. `"sortableAttributes": ["person.firstname"]`

_Much more complete examples are present in the specification_.

Kerollmops commented on text/0000-data-types.md:
```diff
@@ -0,0 +1,319 @@
+        "b": "c",
+    },
+    "a.b": "d"
+}
```
```suggestion
}```
```

gmourier commented on text/0000-data-types.md:
```diff
@@ -0,0 +1,319 @@
+
+#### 2.1.4. Array
+
+An array represents a collection of elements that can be strings or arrays for instance. An array is recursively broken into separate string tokens, which means separate words.
```
📡 Signal @meilisearch/docs-team

gmourier commented:
### Additional questions

@meilisearch/docs-team I've added some examples of how to use dot-notation but I feel this is not the best place to write them.  **See 3.2.4.7. Dot-notation expression**

Do you have a better idea? Some API parameters where it could be used are not specified yet. (running late specification of already existing features)

---

@irevoire See 3.2.4.7. Dot-notation expression section. 

I've added an examples section that demonstrates that the dot-notation can be used to express all properties of an object. 

e.g. ["person"] instead of ["person.firstname", "person.lastname"]. 

This notation could easily work on some parameters while not on others. 

e.g.  `attributesToHighlight` (works-well) vs `searchableAttributes` (impossible, it introduces equals field weight for a subset).

Is it possible to manage this notation for each parameter listed in **3.2.4.7.1. Concerned API parameters section** and report errors in cases where it is not possible?

qdequele commented on text/0121-data-types.md:
````diff
@@ -0,0 +1,450 @@
+        "b": "c",
+    },
+    "a.b": "d"
+}```
````
I think there is an error on the formatting. 
```suggestion
}
```
```

gmourier replied:
Thanks @qdequele. 

This part needs to be decided and precise.

e.g. ["person"] instead of ["person.firstname", "person.lastname"].

This notation could easily work on some parameters while not on others.

e.g. attributesToHighlight (works-well) vs searchableAttributes (impossible, it introduces equals field weight for a subset of document fields).

We could first detect this expression on some parameters and block it with an error message, or simply ignore it and specify that each field must be expressly defined with the notation `.` for a first iteration.

qdequele reviewed (approved):
I've read all the parts about nested fields, including the primitive types. I didn't find anything to complain about. I just wondered about the last example because using `person` to match all the subfields is not explicit in my opinion but it removes a lot of user frustration, so I like it.

gmourier commented:
> I've read all the parts about nested fields, including the primitive types. I didn't find anything to complain about. I just wondered about the last example because using person to match all the subfields is not explicit in my opinion but it removes a lot of user frustration, so I like it.

Thanks @qdequele. 

This part needs to be decided and precise.

e.g. ["person"] instead of ["person.firstname", "person.lastname"].

This notation could easily work on some parameters while not on others.

e.g. attributesToHighlight (works-well) vs searchableAttributes (impossible, it introduces equals field weight for a subset of document fields).

We could first detect this expression on some parameters and block it with an error message, or simply ignore it and specify that each field must be expressly defined with the notation `.` for a first iteration.

qdequele commented:
The problem with doing different behaviors on different routes or resources is that you lose predictability. So it would be nice to have the `person.*` syntax everywhere.

gmourier commented:
What would be the order of importance of the `person` fields in this case according to you? e.g. `searchableAttribute: ["fieldA", "fieldB", "person.*", "fieldD"]`

In that case, I would imagine them naturally defined at the same weight but it requires work that seems out-of-scope for the feature.

One solution for this particular `searchableAttributes` parameter is to have an "undefined" behavior in the same way as `searchableAttributes: ["*"]` which defines the order based on the order of the attributes received the first time.

bidoubiwa commented:
So if I understand correctly, for solution 1 and 3, we use the wildcard in case the user has another field `b` which is not an object?

For solution `2` how would it work? `['a', 'b', 'c']` ?

gmourier commented:
The `.*` notation is now noted as a future possibility, as is `.fi*` or the array notation. e.g. `person.addresses[1]`.

mdubus commented on text/0121-data-types.md:
````diff
@@ -0,0 +1,434 @@
+
+The following JSON document:
+
+```json
+{
+    "a": {
+        "b": "c",
+    },
+    "a.b": "d"
+}
+
+Flattens to:
````
Missing "```" at the end of the `json` example. Sorry I can't make a code suggestion in github for this 🤦‍♀️

Kerollmops commented on text/0121-data-types.md:
````diff
@@ -0,0 +1,435 @@
+```json
+{
+    "attributesToHighlight": [
+        "person.firstname",
````
In case another document have something like `"person": "hello"`.

```suggestion
        "person",
        "person.firstname",
```

Kerollmops commented on text/0121-data-types.md:
```diff
@@ -0,0 +1,435 @@
+    "attributesToHighlight": [
+        "person.firstname",
+        "person.lastname",
+        "person.address.country",
```
```suggestion
        "person.address",
```

dichotommy replied:
```suggestion
        "person.address",
        "person.address.country",
```

dichotommy commented on text/0121-data-types.md:
```diff
@@ -0,0 +1,435 @@
+
+#### 3.2.4.6. searchableAttributes default value case
+
+By default, `searchableAttributes` is set to `[*]`, making all document fields are searchable.
```
```suggestion
By default, `searchableAttributes` is set to `[*]`, making all document fields searchable.
```

dichotommy commented on text/0121-data-types.md:
```diff
@@ -0,0 +1,435 @@
+
+By default, `searchableAttributes` is set to `[*]`, making all document fields are searchable.
+
+In that case, `Attribute` ranking rule consider a field higher in the internal representation more important than a lower one.
```
I believe this is different than what is currently written on the docs. I have opened a PR to adjust accordingly: https://github.com/meilisearch/documentation/pull/1550

dichotommy commented on text/0121-data-types.md:
```diff
@@ -0,0 +1,435 @@
+
+## 4. Future Possibilities
+
+- Change the default behavior of `searchableAttributes` so that it is predictable. We may remove the priority based on a field position in a document.
+- Support the wildcard notation with the dot-notation. e.g. `person.*`, `person.address.*` or `person.l*`
+- Support the array notation. e.g. `person.addresses[1]`
```
A big ➕ 1️⃣ for changing the default behavior of `searchableAttributes`. Right now it seems to be somewhere between defined and undefined behavior (how Meilisearch represents and orders document attributes internally _is_ presumably defined, but we are not sharing that definition w/ the user 😏).

gmourier replied:
Yes, it also bothers me not to be able to explain it clearly or to have to say "undefined behavior". I think we can work on it the day we allow to define the same weight between two fields.

dichotommy reviewed (commented):
I wasn't clear from reading the PR comments, but based on the last three lines of the spec I understand that wildcard dot-notation will not be included in v0.27.

Pretty clear spec, as always! 🧱 🧱 🧱

gmourier commented on text/0121-data-types.md:
````diff
@@ -0,0 +1,435 @@
+}
+```
+
+#### 3.4.1.3. Array of objects mixed with scalar value
````
Yes! This is it

irevoire replied:
I don't know if it's a good idea but it would be fairly easy for us (=the core team) to provide a wasm front end json flatener so people could test it and maybe understand better what is going on.
But doing the front end thingy might take too much time idk

gmourier commented on text/0121-data-types.md:
````diff
@@ -0,0 +1,435 @@
+}
+```
+
+#### 3.2.4.5. Collision between a representation
````
Will double-check, 100% I missed it. Thanks, Tommy!

gmourier commented on text/0121-data-types.md:
````diff
@@ -0,0 +1,435 @@
+
+The following JSON document:
+
+```json
+{
+    "a": [
+        { "b": "c" },
+        { "b": "d" },
+        { "b": "e" },
+    ]
+}
+```
+
+Flattens to:
+
+```json
+{
+    "a.b": ["c", "d", "e"],
+}
+```
````
There is a small difference between your two examples. In the first example, the key are unique per object, the objects are contained within an array.

So we are able to make the value of each searchable object available via `a.b`. 

Imagine that you have an index to find cinemas around you and being to search them by movies title and related genres. He will want to be able to facet and find all the genres rather than the last one. This would be a big limitation and the nested fields would be of little use if Meilisearch only kept the last one in this case.

```json
[
    {
        "id": 0,
        "label": "Pathé Gaumont Paris 13"
        "movies": [
            {
                "title": "Cool Movie 1",
                "genres": ["Summer", "Food", "Buddy"]
            },
            {
                "title": "Wayne's World",
                "genres": ["Stairway To Heaven", "Pepsi"]
            }
        ]
    },
    {
       "id": 1,
       "label": "UGC Paris 13",
       "movies": [
          {
              "title": "Dune",
              "genres": ["Space Opera"]
          }
       ]
    }
]
```

With your first example, the engine will only keep the last movie of the first cinema. It will not work.

While on your second example, it's why the engine does. The fields are located in the same "space" and the last one will be kept.

e.g. Using the last example, if we had a `movie` object containing 2 `genres` attributes

```json
[
    {
        "id": 0,
        "label": "Pathé Gaumont Paris 13"
        "movies": [
            {
                "title": "Cool Movie 1",
                "genres": ["Summer", "Food", "Buddy"]
            },
            {
                "title": "Wayne's World",
                "genres": ["Stairway To Heaven", "Pepsi"]
            }
        ]
    },
    {
       "id": 1,
       "label": "UGC Paris 13",
       "movies": [
          {
              "title": "Dune",
              "genres": ["Space Opera"],
              "genres": ["Sci-fi"]
          }
       ]
    }
]
```

The resulting document in Meilisearch will be

```json
{
       "id": 1,
       "label": "UGC Paris 13",
       "movies": [
          {
              "title": "Dune",
              "genres": ["Sci-fi"]
          }
       ]
}
```

dichotommy replied:
Very helpful, thanks guys!

guimachiavelli commented on text/0121-data-types.md:
```diff
@@ -0,0 +1,435 @@
+
+In the example above, `movie_id`, `1564saqw12ss`, `title`, `Kung fu Panda` are all considered sentences. The colon `:` and comma `,` characters are used as separators.
+
+`"movie_id. 1564saqw12ss. title. Kung fu Panda."`
```
In the paragraph above we say: 

> The colon `:` and comma `,` characters are used as separators.

With that in mind, shouldn't this code sample be:

```suggestion
`"movie_id: 1564saqw12ss. title: Kung fu Panda."`
```

gmourier replied:
Indeed @guimachiavelli, it is not clear.

What it says here is that the keys and values of the document are written this way internally. The `:`, `,` of the JSON document are not kept and the different keys and values are separated by a `.`.

I will modify this section.

guimachiavelli replied:
Ah, I see! Your latest edit makes it much clearer, thanks!

gmourier replied:
@Kerollmops Can you confirm that objects are represented this way internally?

gmourier replied:
@guimachiavelli The spec has been updated!
